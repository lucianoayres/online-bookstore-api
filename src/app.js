import express from 'express'
import { check, param, validationResult } from 'express-validator'
import cors from 'cors'

import clientesRouter from './routes/cliente.route.js'
import autoresRouter from './routes/autor.route.js'
import livrosRouter from './routes/livro.route.js'
import vendasRouter from './routes/venda.route.js'

import basicAuth from 'express-basic-auth'
import { authorizer, authorize } from './controllers/auth.controller.js'

const app = express()
app.use(express.json())
app.use(cors())

app.get('/', async (req, res) => {
  res.status(200).send('Livraria Online API')
})

// all endpoints after this setting will locked under basic auth
app.use(basicAuth({ authorizeAsync: true, authorizer }))

app.use('/cliente', clientesRouter)
app.use('/autor', authorize('admin'), autoresRouter)
app.use('/livro', livrosRouter)
app.use('/venda', vendasRouter)

app.use((err, req, res, next) => {
  res.status(400).send({ error: err.message })
})

export default app
