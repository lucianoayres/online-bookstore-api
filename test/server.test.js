import { jest } from '@jest/globals'
import request from 'supertest'
import app from '../src/app.js'

const adminUser = 'admin'
const adminPassword = 'desafio-igti-nodejs'

const clienteUser = 'ggconceicao@gmail.com'
const clientePassword = 'MRalkmBOJq'

describe('Cenário 01: Testes de Integração', () => {
  /*
  // montagem
  beforeEach(async () => {
    // insert code here
  })

  // desmontagem
  afterAll(async () => {
    //inser code here
  })
  */

  let autorId = null
  let livroId = null
  let livroObj = null
  let clienteId = null
  let clienteLogin = null
  let clienteSenha = null

  jest.setTimeout(300000)
  test('Teste 1: Servidor na porta 3000', async () => {
    const res = await request(app).get('/')
    expect(res.status).toBe(200)
  })

  test('Teste 2: Criar um novo autor válido e verificar se ele foi criado corretamente no banco de dados', async () => {
    // Premissas
    const payloadValid = {
      nome: 'Clarice Lispector',
      email: 'clarisse@gmail.com',
      telefone: '218888-5544'
    }

    // Operação
    const res = await request(app)
      .post('/autor')
      .auth(adminUser, adminPassword)
      .send(payloadValid)

    // Resultado ou Comportamento Esperado
    expect(res.body.erro).toBeUndefined()
    expect(res.status).toBe(201)
    expect(res.body).toMatchSnapshot(payloadValid)
    payloadValid.autorId = autorId = res.body.autorId
    expect(res.body).toMatchObject(payloadValid)
  })

  test('Teste 3: Criar um novo livro válido com dados de teste para o autor criado no teste anterior e verificar se o livro foi criado corretamente', async () => {
    // Premissas
    const payloadValid = {
      nome: 'Serveless Node.js',
      valor: 245.99,
      estoque: 125,
      autorId: autorId
    }

    // Operação
    const res = await request(app)
      .post('/livro')
      .auth(adminUser, adminPassword)
      .send(payloadValid)

    // Resultado ou Comportamento Esperado
    expect(res.body.erro).toBeUndefined()
    expect(res.status).toBe(201)
    expect(res.body).toMatchSnapshot(payloadValid)
    livroObj = res.body
    payloadValid.livroId = livroId = res.body.livroId
    expect(res.body).toMatchObject(payloadValid)
  })

  test('Teste 4: Criar um novo cliente válido com dados de teste e verificar se ele foi criado corretamente', async () => {
    // Premissas
    const payloadValid = {
      nome: 'Jack Blues',
      email: 'jblues@gmail.com',
      senha: 'passpass2021',
      telefone: '819999-8877',
      endereco: 'Rua Gomes da Costa, 892'
    }

    // Operação
    const res = await request(app)
      .post('/cliente')
      .auth(adminUser, adminPassword)
      .send(payloadValid)

    // Resultado ou Comportamento Esperado
    expect(res.body.erro).toBeUndefined()
    expect(res.status).toBe(201)
    expect(res.body).toMatchSnapshot(payloadValid)
    payloadValid.clienteId = clienteId = res.body.clienteId
    clienteLogin = res.body.email
    clienteSenha = res.body.senha
    expect(res.body).toMatchObject(payloadValid)
  })

  test('Teste 5: Buscar o livro criado no teste anterior utilizando dados de login do usário criado no teste anterior e verificar se o retorno é adequado.', async () => {
    jest.setTimeout(100000000)
    // Premissas
    const responseValid = {
      livroId: livroObj.livroId,
      nome: livroObj.nome,
      autorId: livroObj.autorId,
      valor: livroObj.valor,
      estoque: livroObj.estoque
    }
    // Operação
    const res = await request(app)
      .get(`/livro/${livroObj.livroId}`)
      .auth(clienteLogin, clienteSenha)

    // Resultado ou Comportamento Esperado
    expect(res.body.erro).toBeUndefined()
    expect(res.status).toBe(200)

    expect(res.body).toMatchSnapshot(responseValid)
    expect(res.body).toMatchObject(responseValid)
  })

  test('Teste 6: Criar uma venda para o usuário e o livro criados nos testes anteriores e verificar se a venda foi salva corretamente', async () => {
    // Premissas
    const payloadValid = {
      valor: livroObj.valor,
      data: '2021-07-21',
      clienteId: clienteId,
      livroId: livroId
    }
    // Operação
    const res = await request(app)
      .post('/venda')
      .auth(clienteLogin, clienteSenha)
      .send(payloadValid)

    // Resultado ou Comportamento Esperado
    expect(res.body.erro).toBeUndefined()
    expect(res.status).toBe(201)
    //res.body.valor = parseFloat(res.body.valor)
    expect(res.body).toMatchSnapshot(payloadValid)
    payloadValid.vendaId = res.body.vendaId
    expect(res.body).toMatchObject(payloadValid)
  })
})
