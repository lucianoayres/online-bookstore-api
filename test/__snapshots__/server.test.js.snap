// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cenário 01: Testes de Integração Teste 2: Criar um novo autor válido e verificar se ele foi criado corretamente no banco de dados 1`] = `
Object {
  "autorId": 54,
  "email": "clarisse@gmail.com",
  "nome": "Clarice Lispector",
  "telefone": "218888-5544",
}
`;

exports[`Cenário 01: Testes de Integração Teste 3: Criar um novo livro válido com dados de teste para o autor criado no teste anterior e verificar se o livro foi criado corretamente 1`] = `
Object {
  "autorId": 54,
  "estoque": 125,
  "livroId": 56,
  "nome": "Serveless Node.js",
  "valor": 245.99,
}
`;

exports[`Cenário 01: Testes de Integração Teste 4: Criar um novo cliente válido com dados de teste e verificar se ele foi criado corretamente 1`] = `
Object {
  "clienteId": 55,
  "email": "jblues@gmail.com",
  "endereco": "Rua Gomes da Costa, 892",
  "nome": "Jack Blues",
  "senha": "passpass2021",
  "telefone": "819999-8877",
}
`;

exports[`Cenário 01: Testes de Integração Teste 5: Buscar o livro criado no teste anterior utilizando dados de login do usário criado no teste anterior e verificar se o retorno é adequado. 1`] = `
Object {
  "autorId": 54,
  "estoque": 125,
  "livroId": 56,
  "nome": "Serveless Node.js",
  "valor": 245.99,
}
`;

exports[`Cenário 01: Testes de Integração Teste 6: Criar uma venda para o usuário e o livro criados nos testes anteriores e verificar se a venda foi salva corretamente 1`] = `
Object {
  "clienteId": 55,
  "data": "2021-07-21",
  "livroId": 56,
  "valor": 245.99,
  "vendaId": 63,
}
`;
